src/runtime/dsl.ts

```diff
@@ -0,0 +1,431 @@
+          return "uh oh";
+        }
+        if(!found) {
+          found = new DSLVariable(prop);
+          obj.fields[prop] = [found];
+          this.block.registerVariable(found);
+        } else {
+          found = found[0];
+        }
+        return found;
+      },
+      set: (obj:any, prop:string, value:any) => {
+        if(prop === "output") {
+          obj[prop] = value;
+          return true;
+        }
+        if(!obj.fields[prop]) {
+          if(value.constructor !== Array) {
+            value = [value];
+          }
+          obj.fields[prop] = value;
+          return true;
+        }
+        this.block.equivalence(obj.fields[prop][0], value);
+      }
+    })
+  }
+
+  compile() {
+    if(this.output) {
+      return this.toInserts();
+    } else {
+      return this.toScans();
+    }
+  }
+
+  toInserts() {
+    let inserts:(Constraint|Node)[] = [];
+    let e = maybeIntern(this.record.value);
+    let values = [];
+    for(let field in this.fields) {
+      for(let dslValue of this.fields[field]) {
+        let value = this.block.toValue(dslValue) as (RawValue | Register);
+        // @TODO: generate node ids
+        values.push(maybeIntern(value));
+        inserts.push(new InsertNode(e, maybeIntern(field), maybeIntern(value), maybeIntern("my-awesome-node")))
+      }
+    }
+    inserts.push(FunctionConstraint.create("eve/internal/gen-id", {result: e}, values) as FunctionConstraint);
+    return inserts;
+
+  }
+
+  toScans() {
+    let scans:Scan[] = [];
+    let e = maybeIntern(this.record.value);
+    for(let field in this.fields) {
+      for(let dslValue of this.fields[field]) {
+        let value = this.block.toValue(dslValue) as (RawValue | Register);
+        scans.push(new Scan(e, maybeIntern(field), maybeIntern(value), IGNORE_REG))
+      }
+    }
+    return scans;
+  }
+}
+
+//--------------------------------------------------------------------
+// DSLBlock
+//--------------------------------------------------------------------
+
+class DSLBlock {
+  records:DSLRecord[] = [];
+  variables:DSLVariable[] = [];
+  functions:DSLFunction[] = [];
+  variableLookup:{[name:number]:DSLVariable[]} = {};
+  block:runtime.Block;
+
+  constructor(public name:string, public creationFunction:string) {
+    let functionArgs:string[] = [];
+    let code = creationFunction.toString();
+    // trim the function(...) { from the start and capture the arg names
+    code = code.replace(/function\s*\((.*)\)\s*\{/, function(str:string, args:string) {
+      functionArgs.push.apply(functionArgs, args.split(",").map((str) => str.trim()));
+      return "";
+    });
+    // trim the final } since we removed the function bit
+    code = code.substring(0, code.length - 1);
+    code = this.transformBlockCode(code, functionArgs);
+    let neueFunc = new Function(functionArgs[0], functionArgs[1], functionArgs[2], code);
+    neueFunc(this.find, this.record, this.generateLib());
+    this.prepare();
+  }
+
+  generateLib() {
```

Comment:
We find out about this when we try to compile the function, which as far as the user is considered is the same time as the block is declared.